### PR TITLE
Feature/eddi single instance

### PR DESCRIPTION
**Pull request by BullZye54**
Allows only one instance of EDDI system-wide, whether it's started stand-alone or invoked from VoiceAttack. If VoiceAttack is started while EDDI is already running, a MessageBox will be shown to prompt that it can't continue until the existing EDDI instance is closed. Suggestions regarding user interaction are welcome, as now it's mostly for testing.

If VoiceAttack exits while the EDDI config window is open, the EDDI window will be closed.
If the VoiceAttack "configure EDDI" command is spoken while the EDDI window is opened, the command will be ignored. If the window is opened and minimized, it will be restored.

It would be easy to make this a selectable option. As always, feedback and suggestions would be appreciated.

### CODE REVIEW

**Comment by Tkael** [comment]
Nice, though this seems very similar to #322?

**Comment by Tkael** [comment]
Test results: 
Standalone + Standalone: Dialog box appears, 2nd instance doesn't run.
Standalone + VoiceAttack: Dialog appears, after clicking 'ok' VoiceAttack doesn't close and instead the dialog box continues to reappear until the Standalone instance is dismissed.
VoiceAttack + Standalone: Dialog box appears, 2nd instance doesn't run.

**Comment by BullZye54** [comment]
Sorry, I didn't see #322 until after my push request. Yes, similar in function, but implementations do differ, IMO, which I see as a good thing. I see the most single-instance worry to be on the VoiceAttack side, as the startup options for plugins are pretty limited. It does help that VoiceAttack is already single-instance, though.

My next push should clean up some issues with VoiceAttack startup when a stand-alone EDDI instance is active. This is a pretty sticky issue since I can't see a direct way to notify VoiceAttack that the EDDI plugin is available, but not yet ready to process anything. If a VA profile author has set the VA option to run a command when their profile is loaded, and that command uses the EDDI plugin, it becomes especially problematic. Any suggestions on ways to address this issue outside of "don't do that" would be very welcome.

**Comment by BullZye54** [comment]
Aha! I think I might have a slightly less painful way to do this. @Tkael, @richardbuckle, what do think about adding FIFO queue for VA_Invoke1 contexts? If EDDI can't run as a plugin, we could at least bail and show the VA user that the plugin initialization failed and these VA_Invoke1 commands could not be executed.

**Comment by richardbuckle** [comment]
Thanks for this. I'm away right now, hope to be back at my dev machine tomorrow evening when I shall review it. If it's an improvement on #322 I'm all for it!

Question: is VA single-instance per-user or for the entire machine? EDDI only needs to be single-instance per-user FWIW.

**Comment by BullZye54** [comment]
Good question. I haven't tested VA in a multi-user way, nor my code either, for that matter...

**Comment by BullZye54** [comment]
VA is per-user, and so is my single-instance code. It embarrasses me that I didn't think to test that...

**Comment by BullZye54** [comment]
I think it's getting closer, as far as recovery from EDDI instance clashes. I still don't have a solution for when a VA profile author has set the VA option to run a command when their profile is loaded, yet. The plugin has full access to/from VA as far as variables and such are concerned, but if they have any commands that use data provided by other EDDI resources that have not yet started, there could still be issues. However, I don't think it would be any different than what would happen if the plugin encountered an exception while getting the EDDI instance, and the VA user chose to ignore it.

The only thing I've come up with so far is to identify all of the VA_Invoke1 contexts that depend on EDDI-provided data, and issue a warning to VA if those particular providers had an issue. Maybe some sort of context<>provider map?

**Comment by BullZye54** [comment]
I think the instance detection is pretty solid. New VoiceAttack contexts were added:
  * initialize eddi - If the user cancelled the "EDDI instance exists" dialog, and later closes the EDDI window that caused the conflict, an "Initialize eddi" command can be spoken and, if there are no more EDDI instances, and EDDI instance will be fully initialized within VA without it having to be restarted.
 * Added new VA contexts to minimize, maximize, restore and close the EDDI configuration UI. all the contexts can be combined into the existing "Configure EDDI" spoken VA command.

I'm not sure what the policy is for adding VA commands to the EDDI profile, so I've attached a .zip with the modified "Configure eddi" command.

[EDDI-Configure-Command.zip](https://github.com/EDCD/EDDI/files/1609489/EDDI-Configure-Command.zip)

**Comment by BullZye54** [comment]
Defining the VoiceAttack command 'eddi on startup' guarantees that the command will only be run after all EDDI responders and monitors have successfully started. This can allow the user to recover from selecting the "Cancel" option of the EDDI instance conflict message box. If they subsequently close the existing EDDI instance, they can say "Initialize EDDI" to claim the EDDI instance without having to restart VoiceAttack.

**Comment by Tkael** [comment]
Looks pretty good. However, if you cancel initializing EDDI, VoiceAttack's event log still records `4:21:15 PM - Plugin `'EDDI 2.4.6-b3' initialized.` which is incorrect and confusing.
```
4:21:28 PM - The EDDI plugin is not fully initialized.
4:21:27 PM - Recognized : 'configure eddi' (Confidence 67)
4:21:15 PM - Plugin 'EDDI 2.4.6-b3' initialized.
4:21:15 PM - Failed to fully initialise EDDI. Some functions may not work.
```
The `eddi on startup` command is nice, but doesn't match the same convention as other EDDI commands. Revise to match EDDI conventions? e.g. `((EDDI on startup))`, `((EDDI initialized))`, `((EDDI va startup))`?

There's no voice command to unmaximize / unminimize EDDI?

In addition in `Configure EDDI`, perhaps `Open EDDI`?

**Comment by Tkael** [comment]
Pushed an update that I think addresses my comments. 👍

File: EDDI/MainWindow.xaml.cs [hunk 1]
```diff
@@ -329,7 +331,7 @@ private void windowLoaded(object sender, RoutedEventArgs e)
             {
                 if (Properties.Settings.Default.Minimized)
                     senderWindow.WindowState = WindowState.Minimized;
```
Comment (richardbuckle) [inline EDDI/MainWindow.xaml.cs]: Please add braces around all such one-liners. I know it takes up more space but it's a real risk as proven by "goto fail".

Visual Studio can help catch these via its lightbulb tip interface.

File: VoiceAttackResponder/VoiceAttackPlugin.cs [hunk 2]
```diff
@@ -310,93 +359,181 @@ private static void setJsonValues(ref dynamic vaProxy, string prefix, dynamic js
...
-            {
-                Logging.Error("Failed to invoke action " + vaProxy.Context, e);
-                vaProxy.WriteToLog("Failed to invoke action " + vaProxy.Context);
-            }
+                catch (Exception e)
+                {
+                    Logging.Error("Failed to invoke context " + vaProxy.Context, e);
+                    vaProxy.WriteToLog("Failed to invoke context " + vaProxy.Context, "red");
+                }
+            //}
```
Comment (richardbuckle) [inline VoiceAttackResponder/VoiceAttackPlugin.cs]: Commented out close brace will become pollution if left. Please remove.

File: VoiceAttackResponder/VoiceAttackPlugin.cs [hunk 3]
```diff
@@ -310,93 +359,181 @@ private static void setJsonValues(ref dynamic vaProxy, string prefix, dynamic js
...
         }
 
         public static void VA_StopCommand()
         {
         }
 
+        // Allow only one instance of the EDDI configuration UI
+        private static MainWindow configWindow = null;
+        //private static bool configWinRunning = false;
```
Comment (richardbuckle) [inline VoiceAttackResponder/VoiceAttackPlugin.cs]: Commented out code will become pollution if left. Please remove.

**Comment by BullZye54** [comment]
Yep, the "4:21:15 PM - Plugin 'EDDI 2.4.6-b3' initialized" comes from VoiceAttack when a plugin exits the VA_Init1 method, and I can't see a way to suppress it. Given that, I think what's currently being done is about all we can do. Maybe reword the VA log entry from EDDI to replace the word "Fail" with "Unable", which sounds a little less dire? EDDI is technically initialized, as it can still do much of the VA stuff, so it's only "mostly dead" if the responders and monitors fail to start.

- "Open EDDI" it will be, as I like that too. I left it as "Configure" because it was already being used, and I thought that might be the desired convention.

- Say "Restore EDDI" to restore from minimized or maximized states.

- I didn't use the ((EDDI on startup)) format because it's not really a proper "event", but rather a one-time process. And I need to edit ChangeLog.md to remove the "(voice disabled)" annotation because, the more I think about it, it would be more flexible to have it as a normal VA spoken command, so it could be used by both VA and EDDI. I recently read on the VA forum that VA Beta 1.6.9.3 has a "Reload active profile" action, so even "on startup" isn't really the correct phrase, either. I'm open to any and all suggestions on this one. The command is "normally" used to run a process as soon VA startup, but can still be used in certain instances by the user at a later point in time.

Question: what is the procedure normally used put a new/changed VA command into the EDDI profile?

File: VoiceAttackResponder/VoiceAttackPlugin.cs [hunk 4]
```diff
@@ -310,93 +359,181 @@ private static void setJsonValues(ref dynamic vaProxy, string prefix, dynamic js
...
+        private static Thread configThread = null;
+
         private static void InvokeConfiguration(ref dynamic vaProxy)
         {
-            Thread thread = new Thread(() =>
+            string config = (string)vaProxy.Context;
+
+            if (null == configWindow && config != "configuration")
```
Comment (richardbuckle) [inline VoiceAttackResponder/VoiceAttackPlugin.cs]: Yoda syntax distorts readability and is not needed with modern compilers that warn about accidental use of `=` rather than `==`.

Note how it is only used in the first half of the `&&`. Full Yoda syntax would be `if (null == configWindow && "configuration" != config)`.

Note how much more readable the alternative is:
`if (configWindow == null && config != "configuration")`

In any case, if you remembered to use Yoda syntax, you therefore would have remembered to use `==` anyway. 

It is a silly fad that modern compilers have let us do away with, along with "Hungarian" variable prefixes.

**Comment by BullZye54** [comment]
Things are looking pretty good. Thanks, guys, for helping me out!

**Comment by Tkael** [comment]
Thank you for contributing. :-)

**Comment by BullZye54** [comment]
I've hit a regression bug. When the EDDI UI window has been invoked from VoiceAttack, if the window is closed while the "Variables" window is open, it will crash VoiceAttack. This does not happen under EDDI-2.4.6-b2.

I've not had time yet to investigate further. @Tkael, I can't see where a pull requester can update their request status to "hold". Could you please do so until I can get this investigated?

Thanks!

**Comment by richardbuckle** [comment]
I added the Om Hold label for you.

**Comment by BullZye54** [comment]
Sorry about the delay, but I think single-instance is back to a beta-worthy state. When trying to manipulate the EDDI main window from VA, there was a behavior disconnect that happened when the Speech Responder opened a modal dialog/messagebox. The "normal" windows behavior is to disallow interaction with the EDDI window until the modal dialog was closed, but close, minimize, maximize, etc. would still work when sent via the window thread's dispatcher. I could find no MainWindow state variables that suggested that the window was not allowed to receive focus. I'm not terribly WPF-literate, so if y'all know the window properties involved to determine this, I'd sure appreciate it. The solution I have now is a bit of a kludge, but works...

**Comment by Tkael** [comment]
Looks like forward progress on the crash issue, but now the other verbal commands 'Close EDDI', 'Minimize EDDI, 'Maximize EDDI', and 'Restore EDDI' don't seem to be working? Furthermore, the verbal commands 'Configure EDDI' and 'Open EDDI' can be used to open more than one config window.

**Comment by BullZye54** [comment]
Well, this is a real bummer to wake up to. I wouldn't have pushed my changes if I had encountered any problems with the VA integration operations. My testing shows all of what you mention working "as expected", and my first "smoke test" is to ensure that the VA configure/open command allows only one EDDI UI window at a time. Obviously there's something going on, but I currently have no idea as to what the cause may be.

Is there a baseline build/test environment? My VA testing uses the latest "official" VoiceAttack (v1.6.9) and the EDDI.vap profile. Are you using the latest VA beta? And my EDDI build targets the .net Framework 4.5.2, which matches the VS project settings of the repo.

**Comment by Tkael** [comment]
If it's working fully for you, I had better rebuild and try again this evening.  I'm using VA 1.6.9, by the way, and dropping the debug build products into the VA app folder so that I can attach the Visual Studios debugger to VoiceAttack

**Comment by BullZye54** [comment]
Yep, that's been my dev/test method also. EDDI single-instance makes VA integration testing easier because it allows you to block the VA plugin initialize until you've attached the debugger :-).

**Comment by Tkael** [comment]
Ahh... you were blocking the VA plugin until you'd attached the debugger and I wasn't doing that. 
Doing a release build with your build seems to be working quite nicely now, both for single instance and window manipulation.

**Comment by BullZye54** [comment]
A relief to hear that it's behaving better <g>. Thanks!

**Comment by Tkael** [comment]
Hold label removed. @richardbuckle I think this is ready for another look.

**Comment by BullZye54** [comment]
Well, it's still not there. Getting an unhandled exception when swapping ships:
ShipMonitor.cs line 851, shipyard.Add(ship) unhandled exception:

System.NotSupportedException: 'This type of CollectionView does not support changes to its SourceCollection from a thread different from the Dispatcher thread.'

I went back to a clean build of the "dev" branch and EDDI stand-alone works OK, but the same exception was thrown when run as a VA plugin, so it's definately plugin related :-(. I see that this ties into issue #305...

I've been a little time-challenged lately, but I'll try to take a look asap.

**Comment by Tkael** [comment]
Hrm... for the time being would you like this PR to go back on hold?

**Comment by richardbuckle** [comment]
If it also occurs in the development branch then it’s a separate issue and not an obstacle to merging this PR. I think I know what the problem might be: VA calling into our code on a secondary thread, but not being a VA user I can’t prove it. Anyhow, separate issue.

I too have been time-challenged this month but will try to review this PR soon. Also, I am out on the Enigma Expedition, so until the beta starts I won’t be able to test outfitting and shipyard properly. Is it still behaving well in all other respects?

**Comment by Tkael** [comment]
I feel more or less the way. I believe the dispatcher thread is the only link?

**Comment by BullZye54** [comment]
I agree that "the" dispatcher thread is the crux of the matter.

After further testing, I did find a difference between the dev branch and the PR, where the dev branch only failed on a shipyard operation under VA, the PR will fail when stand-alone or from VA if there is a shipyard change detected at ED logon, so it looks like the single-instance startup code change has muddied the waters regarding which is "the" dispatcher thread.

So, a (hopefully short) hold may be in order, for now.

But I think the good news is that when the stand-alone PR fail issue is found, that should certainly help in figuring out what needs to happen when running as a VA plugin. I haven't found any issue with the ship monitor itself as in-game shipyard operations look to be working fine, but the bomb goes off at monitor<->UI update time.

**Comment by BullZye54** [comment]
The ship monitor UI configuration page updates properly when swapping ships, and edits made to the "Role" and "Name (spoken) fields are persisted to the shipmonitor.json file, both stand-alone and plugin.

Unhandled exception crashes when doing shipyard swaps and loadout changes, are now caught, but the cause is still a mystery. ShipMonitor.AddShip and other methods using SynchronizationContext may fail with a "NotSupportedException": "This type of CollectionView does not support changes to its SourceCollection from a thread different from the Dispatcher thread." As an example, buying a ship will result in the Ship monitor UI tab and shipmonitor.json file not getting any of the new ship data.

**Comment by richardbuckle** [comment]
OK thank you but please STOP and RELAX. If we start shovelling more than one fix into a PR then code review gets very hard. Only one subject per PR please!

**Comment by richardbuckle** [comment]
Also, experience has taught us that merging from the `development` branch (which I’m beginning to think should be called the “integration” branch) into a feature branch tends to make the commit history very confusing, often causing a huge amount of duplicated commits. 

So we prefer to keep each feature branch clean and simple: only one topic, for easier review, and no merging in of other branches, for the same reason.

**Comment by BullZye54** [comment]
Sorry for the dev merge, and thanks for the clarification.

I don't intend to add any more changes to the PR. The extra info was meant as an attempt to help point to where the now-caught exception was happening.

File: ShipMonitor/ShipMonitor.cs [hunk 5]
```diff
@@ -27,7 +27,7 @@ public class ShipMonitor : EDDIMonitor
         private int? currentShipId;
         private int? currentProfileId;
         private static readonly object shipyardLock = new object();
-        SynchronizationContext uiSyncContext;
+        //SynchronizationContext uiSyncContext;
```
Comment (richardbuckle) [inline ShipMonitor/ShipMonitor.cs]: We don't like dead code ;)
I'll handle it.

File: SpeechResponder/ConfigurationWindow.xaml.cs [hunk 6]
```diff
@@ -142,6 +144,7 @@ private void testScript(object sender, RoutedEventArgs e)
 
         private void deleteScript(object sender, RoutedEventArgs e)
         {
+            EDDI.Instance.SpeechResponderModalWait = true;
```
Comment (richardbuckle) [inline SpeechResponder/ConfigurationWindow.xaml.cs]: This and subsequent usage is not exception-safe but an acceptable risk for now.

File: EDDI/App.xaml.cs [hunk 7]
```diff
@@ -7,8 +10,26 @@ namespace Eddi
...
+            if (firstOwner)
+            {
+                App app = new App();
+                mainWindow = new MainWindow();
+                app.Run(mainWindow);
+                eddiMutex.ReleaseMutex();
+            }
+            else
+            {
+                MessageBox.Show("An EDDI application instance is already running.",
```
Comment (richardbuckle) [inline EDDI/App.xaml.cs]: I'll see if I can make the copy text more user-oriented.

File: ShipMonitor/ShipMonitor.cs [hunk 8]
```diff
@@ -83,7 +82,7 @@ public UserControl ConfigurationTabItem()
 
         public void Save()
         {
-            uiSyncContext.Post(_ => writeShips(), null);
+            SynchronizationContext.Current.Post(_ => writeShips(), null);
```
Comment (richardbuckle) [inline ShipMonitor/ShipMonitor.cs]: For future reference, this change is bad. You can only retreive the sync context from the main thread. Reinstating my code.

**Comment by Tkael** [comment]
We've got a bug that may be connected to this issue? See https://github.com/EDCD/EDDI/issues/395

**Comment by BullZye54** [comment]
Apologies for the late response. Unfortunately, I've been away from both ED and EDDI for the last several weeks due to other pressing issues. Today was my first exposure to the ED update.

I ran a short test on the new ED and EDDI-Release-3.0.0-b1 where I undocked my ship, flew out of the station, and then auto-docked back in. The VA profile I'm running has VA commands for the "Commander continued", "Undocked" and "Docking granted" EDDI events. On the new release, I did see the described issues where the event either didn't show up in the VA log, or appeared many seconds after the event had triggered.

I then built EDDI-2.4.6-b3 with the single-instance pull request code as submitted and repeated the above, and all three EDDI events showed up in the VA log as the events triggered. When testing while I was working on single-instance, I can't recall any issues with VA *not* picking up the EDDI events as they fired, so at this time, I'm at a bit of a loss as to what the problem may be...

**Comment by Tkael** [comment]
Acknowledged. Thanks for weighing in and testing.

**Comment by BullZye54** [comment]
Could this possibly be a merge clash related to the "Make ShipMonitor events synchronous" (commit ID 4919d22a) changes? The VoiceAttackPlugin VA_Init1 method spins up the updaterThread to maintain VA variables and execute VA commands to trigger on EDDI events, and it's spawned by the VA process. The VA log symptoms suggest that these events may be getting delayed, blocked, or skipped while the EDDI instance is doing its monitor and responder duties? If so, how can/should VA be properly notified that an EDDI event has occurred, across thread boundaries?

And, unfortunately, the pressing (health related) issues I'm currently dealing with have not abated, so I can't yet devote much of any quality time in looking for a potential solution. I am really, really, sorry about this. I will try as best I can...

**Comment by BullZye54** [comment]
Well, the ShipMonitor commit ID I cited doesn't appear to be involved. In looking through the develop branch merge history, there's nothing jumping out at me that might have had an effect on VoiceAttackPlugin, either. But there is definitely something interfering with the operation of the VA plugin worker thread.

**Comment by Tkael** [comment]
It looks like it's the changes we made for status.json. Somehow that's interfering with the VoiceAttack plugin operation. Findings from my investigation so far in #395.

**Comment by Tkael** [comment]
Your health comes first. It looks like you may be on target though... I think the ties to the Instance introduced with status.json may be preventing the threads from operating independently.

**Comment by Tkael** [comment]
Your health comes first. It looks like you may be on target though... I think the ties to the Instance introduced with status.json may be preventing the threads from operating independently.
